src/rules/no-extra-semicolons/index.js

```diff
@@ -0,0 +1,81 @@
+import {
+  ruleMessages,
+  styleSearch,
+  report,
+  validateOptions,
+} from "../../utils"
+
+export const ruleName = "no-extra-semicolons"
+
+export const messages = ruleMessages(ruleName, {
+  rejected: "Unexpected extra semicolon",
+})
+
+function getOffsetByNode(node) {
+  const string = node.root().toString()
+  const nodeColumn = node.source.start.column
+  const nodeLine = node.source.start.line
+  let line = 1
+  let column = 1
+  let index = 0
+
+  for (let i = 0; i < string.length; i++) {
+    if (column === nodeColumn && nodeLine === line) {
+      index = i
+      break
+    }
+
+    if (string[i] === "\n") {
+      column = 1
+      line  += 1
+    } else {
+      column += 1
+    }
+  }
+
+  return index
+}
+
+export default function (actual) {
+  return (root, result) => {
+    const validOptions = validateOptions(result, ruleName, { actual })
+    if (!validOptions) { return }
+
+    if (root.raws.after) {
+      const rawBeforeRoot = root.raws.after
+      styleSearch({ source: rawBeforeRoot, target: ";" }, match => {
+        complain(root.toString().length - rawBeforeRoot.length + match.startIndex)
+      })
+    }
+
+    root.walk((node) => {
+      if (node.raws.before) {
+        const rawBeforeNode = node.raws.before
+        styleSearch({ source: rawBeforeNode, target: ";" }, match => {
+          complain(getOffsetByNode(node) - rawBeforeNode.length + match.startIndex)
```

Comment:
@davidtheclark if line start with `;` and when `\n`, example:
```css
;\na { color: red; }
```
we can `-1 * 2 + 1` = -3 (postcss can't get negative index by function `positionBy`), i can try this, but it is not work. I tried a lot of logic here, this only work that I found.